azure/services/loadbalancers/spec.go

```diff
@@ -0,0 +1,313 @@
+	VNetName             string
+	VNetResourceGroup    string
+	SubnetName           string
+	BackendPoolName      string
+	FrontendIPConfigs    []infrav1.FrontendIP
+	APIServerPort        int32
+	IdleTimeoutInMinutes *int32
+	AdditionalTags       map[string]string
+}
+
+// ResourceName returns the name of the load balancer.
+func (s *LBSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *LBSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for load balancers.
+func (s *LBSpec) OwnerResourceName() string {
+	return ""
+}
+
+// Parameters returns the parameters for the load balancer.
+func (s *LBSpec) Parameters(existing interface{}) (parameters interface{}, err error) {
+	var (
+		etag                *string
+		frontendIDs         []network.SubResource
+		frontendIPConfigs   = make([]network.FrontendIPConfiguration, 0)
+		loadBalancingRules  = make([]network.LoadBalancingRule, 0)
+		backendAddressPools = make([]network.BackendAddressPool, 0)
+		outboundRules       = make([]network.OutboundRule, 0)
+		probes              = make([]network.Probe, 0)
+	)
+
+	if existing != nil {
+		existingLB, ok := existing.(network.LoadBalancer)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.LoadBalancer", existing)
+		}
+		// LB already exists
+		// We append the existing LB etag to the header to ensure we only apply the updates if the LB has not been modified.
+		etag = existingLB.Etag
+		update := false
+
+		// merge existing LB properties with desired properties
+		frontendIPConfigs = *existingLB.FrontendIPConfigurations
+		wantedIPs, wantedFrontendIDs := getFrontendIPConfigs(*s)
+		for _, ip := range wantedIPs {
+			if !ipExists(frontendIPConfigs, ip) {
+				update = true
+				frontendIPConfigs = append(frontendIPConfigs, ip)
+			}
+		}
+
+		loadBalancingRules = *existingLB.LoadBalancingRules
+		for _, rule := range getLoadBalancingRules(*s, wantedFrontendIDs) {
+			if !lbRuleExists(loadBalancingRules, rule) {
+				update = true
+				loadBalancingRules = append(loadBalancingRules, rule)
+			}
+		}
+
+		backendAddressPools = *existingLB.BackendAddressPools
+		for _, pool := range getBackendAddressPools(*s) {
+			if !poolExists(backendAddressPools, pool) {
+				update = true
+				backendAddressPools = append(backendAddressPools, pool)
+			}
+		}
+
+		outboundRules = *existingLB.OutboundRules
+		for _, rule := range getOutboundRules(*s, wantedFrontendIDs) {
+			if !outboundRuleExists(outboundRules, rule) {
+				update = true
+				outboundRules = append(outboundRules, rule)
+			}
+		}
+
+		probes = *existingLB.Probes
+		for _, probe := range getProbes(*s) {
+			if !probeExists(probes, probe) {
+				update = true
+				probes = append(probes, probe)
+			}
+		}
+
+		if !update {
+			// load balancer already exists with all required defaults
+			return nil, nil
+		}
+	}
+
+	frontendIPConfigs, frontendIDs = getFrontendIPConfigs(*s)
```

Comment:
wouldn't this override the existing values if the LB exists and we just want to update it? 

Before the refactor there was a `default` switch statement around this which meant we'd only set these values if existing was nil and it hadn't gone into the above `if existing != nil` logic https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1886/files#diff-85bedcd738722859960dbecef775a3b135f6a27726c7c13005c29b3f07473628L135